src/util/std_expr.h

```diff
@@ -812,6 +812,12 @@ class dereference_exprt : public exprt
     operands().resize(1);
   }
 
+  explicit dereference_exprt(const exprt &op, const typet &tp)
```

Comment:
I don't like having this constructor next to the other one taking only a `typett`. Because they take different `typett`s. This one uses the tp's subtype, the other one does not. This needs at least a comment to warn about the difference, but ideally, this constructor should use its type parameter as the other one does to avoid confusion.